_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+And these often end up blocking meaningful improvements to the safety of users and their data (for
+better or for worse).
+
+In bounds checking, the compiler can see the values involved in the comparisons, as they come from
+constants or other values related to the array. For instance when iterating through a span, stopping
+before the span's `length`, the compiler can see that every index into the span is less than its
+`length` fairly trivially. This usually means the compiler removes the bounds checks. So no runtime
+overhead, yay!
+
+But there's another type of overhead that comes not from the bounds checks in view types like
+`std::span` or `std::string_view`, but from the change to using those types at all.
+
+Here's a commit that converts a bunch of functions to use `base::StringPiece`, the Chromium string
+view type which is much like `std::string_view`:
+[https://chromium-review.googlesource.com/c/chromium/src/+/4575826](
+https://chromium-review.googlesource.com/c/chromium/src/+/4575826).
+
+The functions were previously overloaded to take `const std::string&` or `const char*`. The
+`const char*` overload presumes that:
+- the string being passed in is null terminated, and
+- the function should be consuming the full string up to the null character.
+
+These assumptions are easily wrong when working with view types that encode a bounded view of an
+array. Calling these functions requires converting that bounded string view into a `std::string`
+which involves generating a constructor, and quite possibly a heap allocation. But
+many callers instead convert their bounded string view into a `const char*` through `.data()`
+instead, which is a bug. Props to [@davidben](https://github.com/davidben) for noticing this common
+anti-pattern in Chromium and working to fix them.
+
+So it appears at first glance to be a win to change these functions to take a string view type, in
+this case `base::StringPiece`. For callers with a `std::string`, it will copy the pointer and length
+into the `base::StringPiece`. For callers with a bounded string view type, they don't need to throw
+away the bounds or heap allocate. Yet this created a pretty large binary size regression for doing
+almost nothing, and trying to eliminate heap allocations.
+
+Chromium is an interesting test bed because it is an _enormous_ codebase. So when you make a small
+change, it can easily be amplified immensely. In this case, we discovered that all those calls to a
+`base::StringPiece` constructor grow the binary by 36KB. For unknown reasons, the
+`base::StringPiece` constructor gets inlined, while the `std::string_view` constructor does not, so
+you get less binary size growth with that type (and we'll look at marking `base::StringPiece` with a
+`noinline` attribute).
+
+But in our experience, we've seen that binary size correlates quite well with performance as well.
+So *this also implies a performance degredation by moving from pointers to a bounded view type*. And

```

Comment:
Did we actually observe a regression in time efficiency? Looking at the assembly below, I'd believe it.

Also typo: degradation

Reply:
I did not benchmark these things here, but when we see perf regress we work backward to find the cause by looking where code size increased.

Reply:
Ok I benchmarked and results match up as expected yay. Added links and pictures.